### ISSUE TITLE
HttpBlobStore `get` performance regression

### ISSUE THREAD

**Issue by GregBowyer** [bazelbuild/bazel#4944]
Hi all, in investigating the unexpected performance drop I found after reprogramming the logic in #4889  to use `HttpBlobStore` I happened upon what looks to be a regression in performance.

# Performance analysis
To show this regression, I created two builds (clean, and without AWS S3 work).
The first build is *no-retry*, which is bazel built off 7e50ced9bb59b4ab445edd7904bf31601fd2cea0. This is the commit just before deccc485603c004daad959fd747f1c0c9efc4f00.
The second build is *retry*, which is bazel built off deccc485603c004daad959fd747f1c0c9efc4f00.

Using these builds I can get really basic profiles based on wall clock, these builds are timings on a well connected large build server, building `//src:bazel-bin` from the tree of the relevant version.

To make sure that this is not my terrible code, these are done _without any AWS S3 patches_ and are pointed against google cloud storage. The same bucket is used, but different subpaths. This is done to isolate the two tests as much as possible.

### *no-retry* clean build with an empty cache (`PUT` heavy)
```
INFO: Elapsed time: 157.313s, Critical Path: 70.70s
INFO: Build completed successfully, 2307 total actions

real    2m37.434s
user    0m0.064s
sys     0m0.085s
greg@gregs-desktop ~/projects/bazel-noretry $
```

## *no-retry* clean build, with a full cache (`GET` heavy)
```
INFO: Elapsed time: 25.693s, Critical Path: 16.65s
INFO: Build completed successfully, 2307 total actions

real    0m25.800s
user    0m0.020s
sys     0m0.034s
greg@gregs-desktop ~/projects/bazel-noretry $
```

## *retry* clean build with an empty cache (`PUT` heavy)
```
Target //src:bazel-bin up-to-date:
  bazel-bin/src/bazel
INFO: Elapsed time: 175.435s, Critical Path: 70.29s
INFO: Build completed successfully, 2307 total actions

real    2m55.572s
user    0m0.077s
sys     0m0.089s
greg@gregs-desktop ~/projects/bazel-retry $
```

## *retry* clean build, with a full cache (`GET` heavy)
```
Target //src:bazel-bin up-to-date:
  bazel-bin/src/bazel
INFO: Elapsed time: 1177.332s, Critical Path: 1086.46s
INFO: Build completed successfully, 2307 total actions

real    19m37.448s
user    0m0.232s
sys     0m0.209s
greg@gregs-desktop ~/projects/bazel-retry $
```

# Potential regression site
As can be seen with this commit, the build time shoots up to 19 minutes, slower than not bothering with a cache, and a regression from the previous version of 25seconds.

I think this regression is related to deccc485603c004daad959fd747f1c0c9efc4f00. In this commit it looks like we aim to make it possible to retry operations in the blob store, external from the underlying client (i.e. netty).

Unfortunatly the following code

```java
public void write(int b) throws IOException {
    dataWritten.set(true);
    super.write(b);
}
```

... looks like it makes the download channel expensive. As I see this code in profiler and linux perf output it seems to be making a lot of syscalls into the operating system. This makes the channel handling fairly slow, as it basically makes a syscall to `write()` for _each and every byte_ from the channel.

This also has the additional problem of killing the build parallelism, as anything cache heavy winds up being bound to the two nio threads that are created to service the netty event loop.

# Workaround
A work around is to make the following change, to buffer the output stream for content cache items.

```diff
diff --git a/src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStore.java b/src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStore.java
index bb10623913..361daa61ff 100644
--- a/src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStore.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStore.java
@@ -40,6 +40,7 @@ import io.netty.handler.ssl.SslProvider;
 import io.netty.handler.stream.ChunkedWriteHandler;
 import io.netty.util.internal.PlatformDependent;
 import java.io.ByteArrayInputStream;
+import java.io.BufferedOutputStream;
 import java.io.FileInputStream;
 import java.io.FilterInputStream;
 import java.io.FilterOutputStream;
@@ -189,7 +190,7 @@ public final class HttpBlobStore implements SimpleBlobStore {

   @Override
   public boolean get(String key, OutputStream out) throws IOException, InterruptedException {
-    return get(key, out, true);
+    return get(key, new BufferedOutputStream(out), true);
   }

   @SuppressWarnings("FutureReturnValueIgnored")
@@ -219,6 +220,7 @@ public final class HttpBlobStore implements SimpleBlobStore {
       ch = acquireDownloadChannel();
       ChannelFuture downloadFuture = ch.writeAndFlush(download);
       downloadFuture.sync();
+      out.flush();
       return true;
     } catch (Exception e) {
       // e can be of type HttpException, because Netty uses Unsafe.throwException to re-throw a
```

**Performance with workaround for a clean cached rebuild (`GET` heavy) is:**
```
Target //src:bazel-bin up-to-date:
  bazel-bin/src/bazel
INFO: Elapsed time: 43.799s, Critical Path: 26.75s
INFO: Build completed successfully, 2307 total actions

real    0m43.932s
user    0m0.031s
sys     0m0.035s
greg@gregs-desktop ~/projects/bazel-retry $
```

However I would like to explore if we can use something like
```java
Channel channel = ...;
FileChannel fc = out.getChannel(); // With instanceof cast somewhere *ick*
channel.writeAndFlush(new DefaultFileRegion(fc, 0, fileLength));
```

As that will get us out of touching the file and will better exploit what netty can provide for us.

I dont want to impose on the bazel release schedule, but I think you might want to have this as a blocker for your next release, or maybe do a point release for people caught out by this regression.

Thoughts? comments?

**Comment by ulfjack** [comment]
The correct solution is to fix the FilterOutputStream subclass to do bulk writes. Writing one byte at a time is not going to be fast. I don't think we need to go with the more complex solution of having netty write directly to a file.

**Comment by GregBowyer** [comment]
Humm I am not sure it will wind up being that complicated. However, it might be worth altering FilterOutputStream anyhow for cases where a `FileChannel` cannot be reasonably acquired.

**Comment by ulfjack** [comment]
@buchgr is on vacation for the next week, unfortunately.

I sent out https://bazel-review.googlesource.com/c/bazel/+/48691

**Comment by meteorcloudy** [comment]
I believe this is related to https://github.com/bazelbuild/bazel/issues/4904, I'll try if 0654620 fixes the problem.

**Comment by GregBowyer** [comment]
This still happens for larger artifacts with the changes from @ulfjack. I get the following stacks while handling larger artifacts like `//src:bazel-bin`

```
43 "nioEventLoopGroup-5-2" #680 prio=10 os_prio=0 tid=0x00007f87ac02f800 nid=0x6ec5 runnable [0x00007f8acd8d6000]                                                           
 44    java.lang.Thread.State: RUNNABLE                                                                                                                                      
 45     at java.io.FileOutputStream.write(Native Method)                                                                                                                     
 46     at java.io.FileOutputStream.write(FileOutputStream.java:290)                                                                                                         
 47     at java.io.FilterOutputStream.write(FilterOutputStream.java:77)                                                                                                      
 48     at com.google.devtools.build.lib.remote.blobstore.http.HttpBlobStore$3.write(HttpBlobStore.java:217)                                                                 
 49     at java.io.FilterOutputStream.write(FilterOutputStream.java:125)                                                                                                     
 50     at com.google.devtools.build.lib.remote.blobstore.http.HttpBlobStore$3.write(HttpBlobStore.java:211)                                                                 
 51     at io.netty.buffer.UnsafeByteBufUtil.getBytes(UnsafeByteBufUtil.java:577)                                                                                            
 52     at io.netty.buffer.PooledUnsafeDirectByteBuf.getBytes(PooledUnsafeDirectByteBuf.java:152)                                                                            
 53     at io.netty.buffer.PooledSlicedByteBuf.getBytes(PooledSlicedByteBuf.java:377)                                                                                        
 54     at io.netty.buffer.AbstractByteBuf.readBytes(AbstractByteBuf.java:937)                                                                                               
 55     at com.google.devtools.build.lib.remote.blobstore.http.HttpDownloadHandler.channelRead0(HttpDownloadHandler.java:71)                                                 
 56     at com.google.devtools.build.lib.remote.blobstore.http.HttpDownloadHandler.channelRead0(HttpDownloadHandler.java:39)                                                 
 57     at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)                                                                    
 58     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)                                                          
 59     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)                                                          
 60     at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)                                                            
 61     at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)                              
 62     at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)                                                                        
 63     at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)                                                                            
 64     at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)                                                                  
 65     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)                                                          
 66     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)                                                          
 67     at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)                                                            
 68     at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1379)                                                                                                      
 69     at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1158)                                                                                         
 70     at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1193)                                                                                                      
 71     at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)                                                         
 72     at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428)                                                                             
 73     at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)                                                                            
 74     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)                                                          
 75     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)                                                          
 76     at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)                                                            
 77     at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)                                                                                  
 78     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)                                                          
 79     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)                                                          
 80     at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)                                                            
 81     at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359)                                                                 
 82     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)                                                          
 83     at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)                                                          
 84     at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935)                                                                          
 85     at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:138)                                                                   
 86     at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)                                                                                       
 87     at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)                                                                             
 88     at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)                                                                                      
 89     at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)                                                                                                      
 90     at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)                                                                      
 91     at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)                                                         
 92     at java.lang.Thread.run(Thread.java:748)    
```

**Comment by ulfjack** [comment]
Gah, my bad.

**Comment by GregBowyer** [comment]
No, its fine, it is pretty subtle, let me know when you have a fix in review and I will test again.

**Comment by GregBowyer** [comment]
Also minor question, is the AtomicBoolean used for concurrency or just to handle closing over the boolean.

AFAIK `AtomicBoolean.set` is not atomic ... is this wonky?

**Comment by ulfjack** [comment]
Huh? I am sure that AtomicBoolean.set is atomic.

**Comment by GregBowyer** [comment]
weakly on the atomic, might not be an issue here just an observation

**Comment by ulfjack** [comment]
Can you try this one?
https://bazel-review.googlesource.com/c/bazel/+/48770

**Comment by GregBowyer** [comment]
Is there a way to get diffs from gerrit?

**Comment by GregBowyer** [comment]
nvm found out how to download a patch

**Comment by GregBowyer** [comment]
Winner, there is a small compile fix needed on `flush` but otherwise returns to the kind of performance we want.

```
Target //src:bazel-bin up-to-date:                                                                        
  bazel-bin/src/bazel                                                                                     
INFO: Elapsed time: 21.869s, Critical Path: 13.61s                                                        
INFO: Build completed successfully, 2341 total actions                                                    
                                                                                                          
real    0m21.891s                                                                                         
user    0m0.018s                                                                                          
sys     0m0.033s                                                                                          
greg@gregs-desktop ~/projects/bazel-retry $
```

**Comment by GregBowyer** [comment]
As for the volatile, reading the code a bit more I *think* its ok it should be a weak barrier between the two threads. Ignore that concern.

**Comment by meteorcloudy** [comment]
I can also confirm that https://bazel-review.googlesource.com/c/bazel/+/48770 fixed #4904

**Comment by ulfjack** [comment]
Great, thanks for the confirmation, both of you. I am in the process of committing the rollback, and I just sent out the new fix for review.
